### PR TITLE
8340401: DcmdMBeanPermissionsTest.java and SystemDumpMapTest.java fail with assert(_stack_base != nullptr) failed: Sanity check

### PR DESCRIPTION
**Pull request by stooke**
This PR addresses JDK-8340401 by adding a test for a thread with no stack currently assigned.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8340401](https://bugs.openjdk.org/browse/JDK-8340401): DcmdMBeanPermissionsTest.java and SystemDumpMapTest.java fail with assert(_stack_base != nullptr) failed: Sanity check (**Bug** - P3)

### Reviewers
 * [Thomas Stuefe](https://openjdk.org/census#stuefe) (@tstuefe - **Reviewer**) Review applies to [4e4f0fe3](https://git.openjdk.org/jdk/pull/21091/files/4e4f0fe334281fca52a692f74553f11df05c5419)
 * [Kevin Walls](https://openjdk.org/census#kevinw) (@kevinjwalls - **Reviewer**) Review applies to [f8f8bbab](https://git.openjdk.org/jdk/pull/21091/files/f8f8bbabbb9aa0c2bddfc14a5fbe3d478ed022b7)
 * [David Holmes](https://openjdk.org/census#dholmes) (@dholmes-ora - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/21091/head:pull/21091` \
`$ git checkout pull/21091`

Update a local copy of the PR: \
`$ git checkout pull/21091` \
`$ git pull https://git.openjdk.org/jdk.git pull/21091/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 21091`

View PR using the GUI difftool: \
`$ git pr show -t 21091`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/21091.diff">https://git.openjdk.org/jdk/pull/21091.diff</a>

</details>
<details><summary>Using Webrev</summary>

[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/21091#issuecomment-2363068188)
</details>

### CODE REVIEW

**Comment by kevinjwalls** [comment]
Would be good to go ahead with this.
It seems on Windows that JavaThread with undefined stack base/size can be seen via JavaThreadIteratorWithHandle.
There may be more to investigate as to why this is happening, but guarding against it as a quick fix seems good as the failure is quite noisy.
I'm testing it too, will let you know if anything interesting happens.

**Comment by kevinjwalls** [comment]
Nothing interesting happened. Meaning, no crashes!

File: src/hotspot/share/nmt/memMapPrinter.cpp
Comment (dholmes-ora) [inline src/hotspot/share/nmt/memMapPrinter.cpp:176]: I think this should go in the iteration loop:
```
  for (JavaThreadIteratorWithHandle jtiwh; JavaThread* t = jtiwh.next(); ) {
    // We can encounter new threads whose stack has not yet been initialized - so skip them
    if (t->stack_size() == 0) continue;
    HANDLE_THREAD(t);
  }
```
or even
```
  for (JavaThreadIteratorWithHandle jtiwh; JavaThread* t = jtiwh.next(); ) {
    // We can encounter new threads whose stack has not yet been initialized - so skip them in debug to avoid assertion
    DEBUG_ONLY(if (t->stack_size() == 0) continue;)
    HANDLE_THREAD(t);
  }
```
this assumes `range_intersects` will correctly handle a zero base and end.

Comment (tstuefe) [reply src/hotspot/share/nmt/memMapPrinter.cpp:176]: I would prefer it here, or in HANDLE_THREAD macro, to make sure we do this for all threads printed.

Comment (stooke) [reply src/hotspot/share/nmt/memMapPrinter.cpp:176]: I will put it in HANDLE_THREAD to cover the other tests.

**Review by tstuefe — COMMENTED** [review]
I don't think this is good enough. The setting of stack size can race the setting of stack_base - you could observe stack size set, but stack base still unset.

What I would do instead is:

give us a new function in Thread that does not assert (similar to e.g. `Thread::curren_or_null()`):

```
address stack_base_or_null() const { return _stack_base; }
```
use this in `vma_touches_thread_stack` to skip NULL base.

Alternatively, in `record_stack_base_and_size`, we need a store store barrier between the two writes, and store stack size first.

**Comment by tstuefe** [comment]
> Would be good to go ahead with this. It seems on Windows that JavaThread with undefined stack base/size can be seen via JavaThreadIteratorWithHandle. There may be more to investigate as to why this is happening, but guarding against it as a quick fix seems good as the failure is quite noisy. I'm testing it too, will let you know if anything interesting happens.

I think we have understood why this is happening on Windows. See David's comment in the JBS issue.

File: src/hotspot/share/nmt/memMapPrinter.cpp
Comment (dholmes-ora) [reply src/hotspot/share/nmt/memMapPrinter.cpp]: It is not needed for the other tests - those threads can't be new.

This is an issue with using the JavaThread iterator so should be visibly seen in the itertation loop IMO.

**Comment by dholmes-ora** [comment]
> Alternatively, in record_stack_base_and_size, we need a store store barrier between the two writes, and store stack size first.

I guess this could be a real issue on Windows-aarch64. The storestore barrier would work for me. But we probably need an audit of every JavaThread iterator use in case the same problem can occur elsewhere.

**Comment by tstuefe** [comment]
> I tested this update in a local tier1 linux and windows, and in the CI tier3 and tier5 windows where I'm seeing all passes, lots of ZGC usage (which remains a more likely way to trigger this, but not essential).

Have you tested with Windows arm, too?

> 
> Simon if this attracts another reviewer, I would be happy to see this go ahead and calm down the CI noise (every tier5 run turns up a few failures).
> 
> Like the other test change, if we want to meditate further on it and follow up with a better solution that's great.

Can't we just problem-list it for now? I would like that better than adding a solution that we already know won't work for all cases (which would make e.g. hunting the bug tail when backporting more confusing).

**Comment by kevinjwalls** [comment]
> Have you tested with Windows arm, too?

No.  I get the theory, but has anybody seen an issue there?
I'm not against problemlisting if another real solution is in the works.

**Comment by tstuefe** [comment]
> > Have you tested with Windows arm, too?
> 
> No. I get the theory, but has anybody seen an issue there? I'm not against problemlisting if another real solution is in the works.

Yes, Simon is working on a solution. This may just be a case of too-many-reviewer-opinions-paralyses. Can you problemlist the test or should we do it?

**Comment by kevinjwalls** [comment]
> > > Have you tested with Windows arm, too?
> > 
> > 
> > No. I get the theory, but has anybody seen an issue there? I'm not against problemlisting if another real solution is in the works.
> 
> Yes, Simon is working on a solution. This may just be a case of too-many-reviewer-opinions-paralyses. Can you problemlist the test or should we do it?

Would be great if you could, thanks.

**Comment by tstuefe** [comment]
> > > > Have you tested with Windows arm, too?
> > > 
> > > 
> > > No. I get the theory, but has anybody seen an issue there? I'm not against problemlisting if another real solution is in the works.
> > 
> > 
> > Yes, Simon is working on a solution. This may just be a case of too-many-reviewer-opinions-paralyses. Can you problemlist the test or should we do it?
> 
> Would be great if you could, thanks.

okay: https://github.com/openjdk/jdk/pull/21109

**Comment by dholmes-ora** [comment]
> I don't think this is limited to JavaThread

@tstuefe whether other kinds of thread iterators have potential issues with the state of the threads they may observe is not relevant to the question of whether the proposed check goes in the iteration loop or the HANDLE_THREAD macro - we don't process any GC threads with HANDLE_THREAD. The JavaThread iterator can return a thread that has not yet set its stack so the code using the iterator needs to deal with that if it might be an issue.

**Comment by dholmes-ora** [comment]
> Yes, Simon is working on a solution. 

On solution may be to stop creating the threads in the suspended state and use the same protocol for initialization that is used for non-Windows threads. At least that way we get consistent behaviour. I wouldn't expect Simon to undertake that task though.

**Comment by tstuefe** [comment]
> > I don't think this is limited to JavaThread
> 
> @tstuefe whether other kinds of thread iterators have potential issues with the state of the threads they may observe is not relevant to the question of whether the proposed check goes in the iteration loop or the HANDLE_THREAD macro - we don't process any GC threads with HANDLE_THREAD. The JavaThread iterator can return a thread that has not yet set its stack so the code using the iterator needs to deal with that if it might be an issue.

We iterate GC threads (https://github.com/openjdk/jdk/blob/9b54225e35da5411e7a9e51320adc3b014b92261/src/hotspot/share/nmt/memMapPrinter.cpp#L226) here; it calls vma_touches_thread_stack, which is why I proposed in my last comment (https://github.com/openjdk/jdk/pull/21091#pullrequestreview-2317615786) to fix that function. Am I overlooking something?

**Comment by dholmes-ora** [comment]
@tstuefe I think this is a thread lifecycle management issue and so anyone using the JavaThread iterator or GCworkersDo needs to be aware of the possibility of a thread having a null stackbase. Having the check as close as possible to the iteration loop (whichever kind) makes this more visible, rather than hiding it inside another function. I would rather `vma_touches_thread_stack` be guaranteed to get "good threads" than writing it to allow for "bad ones". Of course even better if we simply don't allow this kind of thread "badness" in the first place.

**Comment by tstuefe** [comment]
> @tstuefe I think this is a thread lifecycle management issue and so anyone using the JavaThread iterator or GCworkersDo needs to be aware of the possibility of a thread having a null stackbase. Having the check as close as possible to the iteration loop (whichever kind) makes this more visible, rather than hiding it inside another function. I would rather `vma_touches_thread_stack` be guaranteed to get "good threads" than writing it to allow for "bad ones". Of course even better if we simply don't allow this kind of thread "badness" in the first place.

Okay, that makes sense considering that people tend to copy paste solutions. 

I leave this up to Simon, then (also the question whether he wants to tackle rewriting Windows thread startup in a manner that we prevent this situation altogether). After all, the test now is problemlisted, so the urgency is gone.

**Comment by dholmes-ora** [comment]
As I've heard nothing further here, I've filed [JDK-8342482](https://bugs.openjdk.org/browse/JDK-8342482) and will work on it.

**Comment by dholmes-ora** [comment]
[JDK-8342482](https://bugs.openjdk.org/browse/JDK-8342482) didn't work out.

**Comment by kevinjwalls** [comment]
Hi @stooke Simon, do you want to progress this change or some version of it?
If so, removing the test from the problemlist would be done at the same time.
To get into jdk24, it would need to happen in the next days or so... 8-)

**Review by tstuefe — COMMENTED** [review]
I still would prefer for vma_touches_thread_stack to be made more robust instead, since it  is used also for iteration of GC threads. Logically, it should return null for a `[null, null+0)` stack input range.

**Comment by stooke** [comment]
@tstuefe I have created a branch with my proposed fix, with a more robust vma_touches_thread_stack() using the new Thread::stack_base_or_null()  method.  It's at [this branch](https://github.com/stooke/jdk/tree/pr_windows_system_dump_fixes2)  I will merge it back here if you approve of the changes.

Should I un-problemlist the tests in this PR, or should that be a separate PR?

**Comment by tstuefe** [comment]
@stooke Yes, I like that better, safe I don't believe we need `Thread::stack_base_or_null` for just one use case. I would do the t->stack_base != NULL test in vma_touches_thread_stack before the range compare (I also keep thinking we should have a generic Range class for this and eventually convert the zillion (start, end) ... into uses of that Range class, and such a class would almost certainly barf at negative ranges (start > end)

**Comment by stooke** [comment]
@tstuefe  we can't call stack_base() because it will throw an assert if it's null. (hence the stack_base_or_null() addition). Would you like me to check stack_size() instead? (my suggestion is to rename stack_base_or_null() to a more generic "has_valid_stack()")

**Comment by tstuefe** [comment]
> @tstuefe we can't call stack_base() because it will throw an assert if it's null. (hence the stack_base_or_null() addition). Would you like me to check stack_size() instead? (my suggestion is to rename stack_base_or_null() to a more generic "has_valid_stack()")

Oh you are right, I forgot that. No, I think stack_base_or_null is fine in that case.

**Comment by stooke** [comment]
This was closed in error.

**Comment by stooke** [comment]
@tstuefe I have implemented Thread::stack_base_or_null(), and I expect the Gitlab gates to pass.  Two questions:
1) should I check in a SystemDumpMapZGCTest() ?
2) should I remove these tests from the ProblemList files?

**Review by tstuefe — APPROVED** [review]
Good.

**Comment by dholmes-ora** [comment]
@stooke if there is any test in the ProblemList file that lists 8340401, then the PL needs to be updated. Thanks

File: src/hotspot/share/runtime/thread.hpp
Comment (dholmes-ora) [inline src/hotspot/share/runtime/thread.hpp:528]:
```suggestion
  // Needed for code that can query a new thread before the stack has been set.
  address stack_base_or_null() const   { return _stack_base; }
```

File: src/hotspot/share/nmt/memMapPrinter.cpp
Comment (dholmes-ora) [inline src/hotspot/share/nmt/memMapPrinter.cpp:175]: Possibly add a comment:
```
// Note it is possible to encounter a brand new thread that has not yet initialized its stack fields.
```

File: src/hotspot/share/runtime/thread.hpp
Comment (stooke) [reply src/hotspot/share/runtime/thread.hpp:528]: Done.

File: src/hotspot/share/nmt/memMapPrinter.cpp
Comment (stooke) [reply src/hotspot/share/nmt/memMapPrinter.cpp:175]: Done.

**Comment by stooke** [comment]
> @stooke if there is any test in the ProblemList file that lists 8340401, then the PL needs to be updated. Thanks

Done.

**Review by dholmes-ora — APPROVED** [review]
Thanks for updates

**Comment by dholmes-ora** [comment]
SystemDumpMapTest still fails after this - see [JDK-8346717](https://bugs.openjdk.org/browse/JDK-8346717)

**Comment by stooke** [comment]
> SystemDumpMapTest still fails after this - see [JDK-8346717](https://bugs.openjdk.org/browse/JDK-8346717)

@dholmes-ora my apologies - this failure (which is timing related) did not sho up in my testing.  I will be opening PR #22870

**Pull request closed without merging**
